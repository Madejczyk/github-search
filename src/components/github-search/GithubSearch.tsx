import * as React from "react";
import { SearchInput } from "./SearchInput";
import { SubmitButton } from "./SubmitButton";
import { Title } from "./Title";
import { UserDetails, UserDetailsProps } from "./UserDetails";

enum LoadingState {
    NOT_LOADED = "NOT_LOADED",
    LOADING = "LOADING",
    LOADED = "LOADED",
    ERROR = "ERROR",
}

enum SubmitButtonValue {
    NOT_LOADED = "Search",
    LOADING = "Loading ...",
    LOADED = "Sucess :)",
    ERROR = "Error :(",
}

type GithubSearchState = {
    errorMessage: string,
    loadingState: LoadingState,
    userInfo?: {
        avatar_url: string
        bio: string | null,
        name: string,
        repos_url: string,
    }
    userName: string,
};

const USERNAME_REGEX = /^[a-z\d](?:[a-z\d]|-(?=[a-z\d])){0,38}$/i;

export class GithubSearch extends React.PureComponent<{}, GithubSearchState> {
    constructor(props: {}) {
        super(props);
        this.state = {
            errorMessage: "",
            loadingState: LoadingState.NOT_LOADED,
            userInfo: undefined,
            userName: "",
        };
        this.onChange = this.onChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    private onChange(event: React.ChangeEvent<HTMLInputElement>) {
        this.setState({userName: event.target.value});
    }

    private handleSubmit(event: React.FormEvent<HTMLFormElement>) {
        this.setState({loadingState: LoadingState.LOADING});
        event.preventDefault();
        this.findUserDetails();
    }

    private async findUserDetails() {
        const url = `https://api.github.com/users/${this.state.userName}`;
        const response = await fetch(url);
        if (!response.ok) {
            this.setState({
                errorMessage: response.statusText,
                loadingState: LoadingState.ERROR,
            });
        } else {
            const result = await response.json();
            const {name, bio, avatar_url, repos_url} = result;
            this.setState({
                loadingState: LoadingState.LOADED,
                userInfo: {
                    avatar_url,
                    bio,
                    name,
                    repos_url,
                },
            });
        }
    }

    private isSubmitButtonDisabled = () => {
        return !USERNAME_REGEX.test(this.state.userName) || this.state.loadingState === LoadingState.LOADING;
    }

    private getSubmitButtonMessage = () => {
        return this.state.loadingState === LoadingState.ERROR ?
            this.state.errorMessage : SubmitButtonValue[this.state.loadingState];
    }

    public render() {
        return (
            <form onSubmit={this.handleSubmit} className="github-search">
                <Title/>
                <SearchInput
                    onChange={this.onChange}
                    value={this.state.userName}
                />
                <SubmitButton
                    isDisabled={this.isSubmitButtonDisabled()}
                    text={this.getSubmitButtonMessage()}
                />
                {
                    this.state.userInfo && <UserDetails
                        userInfo={this.state.userInfo}
                    />
                }
            </form>
        );
    }
}
