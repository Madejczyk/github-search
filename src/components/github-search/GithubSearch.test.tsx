import { mount } from "enzyme";
import * as React from "react";
import { GithubSearch } from "./GithubSearch";

describe("GithubSearch", () => {
    test("should initial contains disabled search button", () => {
        const component = mount(<GithubSearch />);
        expect(component.state()).toEqual({
            loadingState: "NOT_LOADED",
            userName: "",
        });
        expect(component.find(".submit-button").prop("disabled")).toBe(true);
        expect(component.render()).toMatchSnapshot();
    });

    test("should has enable search button when username is changed", () => {
        const component = mount(<GithubSearch />);
        component.find(".search-input").simulate("focus");
        component.find(".search-input").simulate("change", {
            target: {
                value: "test",
            },
        });
        expect(component.state()).toEqual({
            loadingState: "NOT_LOADED",
            userName: "test",
        });
        expect(component.find(".submit-button").prop("disabled")).toBe(false);
        expect(component.render()).toMatchSnapshot();
    });
});
