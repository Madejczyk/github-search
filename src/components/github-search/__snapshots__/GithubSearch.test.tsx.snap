// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GithubSearch should has enable search button when username is changed 1`] = `
<form
  class="github-search"
>
  <div
    class="title"
  >
    GitHub User details
  </div>
  <input
    class="search-input"
    placeholder="User name"
    type="text"
    value="madejczyk"
  />
  <input
    class="submit-button"
    type="submit"
    value="Search"
  />
</form>
`;

exports[`GithubSearch should initial contains disabled search button 1`] = `
<form
  class="github-search"
>
  <div
    class="title"
  >
    GitHub User details
  </div>
  <input
    class="search-input"
    placeholder="User name"
    type="text"
    value=""
  />
  <input
    class="submit-button"
    disabled=""
    type="submit"
    value="Search"
  />
</form>
`;

exports[`GithubSearch should set error when api returns error 1`] = `
<form
  class="github-search"
>
  <div
    class="title"
  >
    GitHub User details
  </div>
  <input
    class="search-input"
    placeholder="User name"
    type="text"
    value="madejczy"
  />
  <input
    class="submit-button"
    type="submit"
    value="Not Found"
  />
</form>
`;

exports[`GithubSearch should set loaded status when username is correct 1`] = `
<form
  class="github-search"
>
  <div
    class="title"
  >
    GitHub User details
  </div>
  <input
    class="search-input"
    placeholder="User name"
    type="text"
    value="madejczyk"
  />
  <input
    class="submit-button"
    type="submit"
    value="Sucess :)"
  />
</form>
`;

exports[`GithubSearch should set loading status after submit 1`] = `
<form
  class="github-search"
>
  <div
    class="title"
  >
    GitHub User details
  </div>
  <input
    class="search-input"
    placeholder="User name"
    type="text"
    value="madejczyk"
  />
  <input
    class="submit-button"
    disabled=""
    type="submit"
    value="Loading ..."
  />
</form>
`;
